wooting-macro-backend/src/config.rs

```diff
@@ -58,85 +63,121 @@ pub trait ConfigFile: Default + serde::Serialize + for<'de> serde::Deserialize<'
     /// Writes the config file to the config directory.
     ///
     /// If it fails, it uses only in-memory defaults and won't save anything to disk.
-    fn write_to_file(&self) {
-        let converted_json = match serde_json::to_string_pretty(&self) {
-            Ok(value) => value,
-            Err(err) => {
-                error!(
-                    "Error converting data from json to types, aborting write.\n{}",
-                    err
-                );
-                return;
-            }
-        };
+    fn write_to_file(&self) -> Result<()> {
+        let converted_json = serde_json::to_string_pretty(&self)?;
 
-        match std::fs::write(Self::file_name().as_path(), converted_json) {
-            Ok(_) => {
-                info!("Success writing a new file");
-            }
-            Err(err) => {
-                error!("Error writing a new file. {}", err);
-            }
-        };
+        std::fs::write(Self::file_name()?.as_path(), converted_json)?;
+
+        Ok(())
     }
 }
 
 impl ConfigFile for ApplicationConfig {
-    fn file_name() -> PathBuf {
+    fn file_name() -> Result<PathBuf> {
+        let path: PathBuf;
         let dir = {
             #[cfg(debug_assertions)]
-            let x = PathBuf::from("..");
+            {
+                path = PathBuf::from("..");
+            }
 
             #[cfg(not(debug_assertions))]
-            let x = dirs::config_dir().unwrap().join(CONFIG_DIR);
+            {
+                path = {
+                    let conf_dir: Result<PathBuf> = match dirs::config_dir() {
+                        Some(config_path) => Ok(config_path),
+                        None => Err(anyhow::Error::msg(
+                            "Cannot find config directory, cannot proceed.",
+                        )),
+                    };
+                    conf_dir?.join(CONFIG_DIR)
+                };
+            }
 
-            x
+            path
         };
 
-        dir.join(CONFIG_FILE)
+        Ok(dir.join(CONFIG_FILE))
     }
 }
 
 impl ConfigFile for LogDirPath {
-    fn file_name() -> PathBuf {
+    fn file_name() -> Result<PathBuf> {
+        let path: PathBuf;
         #[cfg(debug_assertions)]
-        let x = PathBuf::from("..");
+        {
+            path = PathBuf::from("..");
+        }
 
         #[cfg(not(debug_assertions))]
-        let x = dirs::config_dir().unwrap().join(CONFIG_DIR);
+        {
+            path = {
+                let conf_dir: Result<PathBuf> = match dirs::config_dir() {
+                    Some(config_path) => Ok(config_path),
+                    None => Err(anyhow::Error::msg(
+                        "Cannot find config directory, cannot proceed.",
+                    )),
+                };
+                conf_dir?.join(CONFIG_DIR)
+            };
+        }
```

Comment:
What's the real world situation in where the config_dir will fail to retrieve? This seems more like an unwrap and panic situation than bubbling up an error

Reply:
What happens when it panics? Cus in some cases you should if it's a hard error and be able to determine the panic state

Reply:
> What happens when it panics? Cus in some cases you should if it's a hard error and be able to determine the panic state

It should bubble up and then it's decided by the upper functions what to do. I think i made it panic in the end in ``src-tauri``, where it's used most.

Reply:
![image](https://github.com/WootingKb/wooting-macros/assets/1900179/4b474210-1460-4d9b-8c60-ca1e8868fd93)

![image](https://github.com/WootingKb/wooting-macros/assets/1900179/8e6eb38d-e969-4935-84ab-d342c5108a4f)

Yeah, these errors get .expected later on, it's just there so the caller can figure out whether to panic outright. Not sure if the best approach but figured it's a better one than just randomly panicking somewhere deeper in library. 

We don't really handle cases where there's no XDG directories setup, we just panic.